aws/resource_aws_appmesh_virtual_router.go

```diff
@@ -0,0 +1,159 @@
+		Create: resourceAwsAppmeshVirtualRouterCreate,
+		Read:   resourceAwsAppmeshVirtualRouterRead,
+		Update: resourceAwsAppmeshVirtualRouterUpdate,
+		Delete: resourceAwsAppmeshVirtualRouterDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"mesh_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"spec": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"service_names": {
+							Type:     schema.TypeSet,
+							Required: true,
+							MinItems: 1,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+					},
+				},
+			},
+
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"created_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"last_updated_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAppmeshVirtualRouterCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appmeshconn
+
+	req := &appmesh.CreateVirtualRouterInput{
+		MeshName:          aws.String(d.Get("mesh_name").(string)),
+		VirtualRouterName: aws.String(d.Get("name").(string)),
+		Spec:              expandAppmeshVirtualRouterSpec(d.Get("spec").([]interface{})),
+	}
+
+	log.Printf("[DEBUG] Creating App Mesh virtual router: %#v", req)
+	resp, err := conn.CreateVirtualRouter(req)
+	if err != nil {
+		return fmt.Errorf("error creating App Mesh virtual router: %s", err)
+	}
+
+	d.SetId(aws.StringValue(resp.VirtualRouter.Metadata.Uid))
+
+	return resourceAwsAppmeshVirtualRouterUpdate(d, meta)
+}
+
+func resourceAwsAppmeshVirtualRouterUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appmeshconn
+
+	if !d.IsNewResource() && d.HasChange("spec") {
+		_, v := d.GetChange("spec")
+		req := &appmesh.UpdateVirtualRouterInput{
+			MeshName:          aws.String(d.Get("mesh_name").(string)),
+			VirtualRouterName: aws.String(d.Get("name").(string)),
+			Spec:              expandAppmeshVirtualRouterSpec(v.([]interface{})),
+		}
+
+		log.Printf("[DEBUG] Updating App Mesh virtual router: %#v", req)
+		_, err := conn.UpdateVirtualRouter(req)
+		if err != nil {
+			return fmt.Errorf("error updating App Mesh virtual router: %s", err)
+		}
+	}
+
+	return resourceAwsAppmeshVirtualRouterRead(d, meta)
+}
+
+func resourceAwsAppmeshVirtualRouterRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appmeshconn
+
+	resp, err := conn.DescribeVirtualRouter(&appmesh.DescribeVirtualRouterInput{
+		MeshName:          aws.String(d.Get("mesh_name").(string)),
+		VirtualRouterName: aws.String(d.Get("name").(string)),
+	})
+	if err != nil {
+		if isAWSErr(err, "NotFoundException", "") {
```

Comment:
Nit: SDK provided constant available:

```suggestion
		if isAWSErr(err, appmesh.ErrCodeNotFoundException, "") {
```